sympy/solvers/ode.py

```diff
@@ -363,25 +362,15 @@ def sub_func_doit(eq, func, new):
     x*(-1/(x**2*(z + 1/x)) + 1/(x**3*(z + 1/x)**2)) + 1/(x*(z + 1/x))
     ...- 1/(x**2*(z + 1/x)**2)
     """
-    reps = {}
-    repu = {}
-    for d in eq.atoms(Derivative):
-        u = Dummy('u')
-        repu[u] = d.subs(func, new).doit()
-        reps[d] = u
-
-    # Make sure that expressions such as ``Derivative(f(x), (x, 2))`` get
-    # replaced before ``Derivative(f(x), x)``:
-    #
-    # Also replace e.g. Derivative(x*Derivative(f(x), x), x) before
-    # Derivative(f(x), x)
-    def cmp(subs1, subs2):
-        return subs2[0].has(subs1[0]) - subs1[0].has(subs2[0])
-    key = lambda x: (-x[0].derivative_count, cmp_to_key(cmp)(x))
-    reps = sorted(reps.items(), key=key)
-
-    return eq.subs(reps).subs(func, new).subs(repu)
-
+    reps = {func: new}
+    if order:
+        x = func.args[0]
+        for i in range(order, -1, -1):
+            reps[func.diff(x, i)] = new.diff(x, i)
+    else:
+        for d in eq.atoms(Derivative):
+            reps[d] = d.subs(func, new).doit(deep=False)
```

Comment:
the only other thing that might make this slow in comparison to the above is that this is using `d.subs` rather than computing the derivative directly...maybe the subs is slow and if the rhs. But now another complicating factor occurs to me: any given derivative might not be constant-free, e.g. `Derivative(y*f(x),x).subs(f(x).diff(x),x)` in which case the above would fail.

So perhaps:
```python
reps= {func: new}
for d in eq.atoms(Derivative):
    if d.expr == func:
        reps[d] = new.diff(*d.variable_count)
    else:
        reps[d] = d.xreplace({func: new}).doit(deep=False)
return eq.xreplace(reps)
```

Reply:
a test of that complicating issue should be added

Reply:
This is not necessary since `_preprocess` already makes sure derivatives are evaluated.